Initial overlay example for Keda and Structure pre-commit hook (#65897)
mistercrunch/Airflow: chart/kustomize-overlays/CONTRIBUTING.rst

   KIND, either express or implied.  See the License for the
   specific language governing permissions and limitations
   under the License.

Contributing Kustomize Overlays
===============================

This document is the authoritative reference for adding, evolving, and
retiring overlays under ``chart/kustomize-overlays/``.

Why this directory exists
-------------------------

The Airflow Helm chart has historically carried components that are not
Airflow-native. They make the chart heavier than it needs to be and pull
maintenance toward things that already have external owners. Expressing
these components as Kustomize overlays keeps the chart focused on Airflow
itself while still giving users a working starting point for the rest.

The chart never removes a component without a working overlay already in
place. Users always have a migration path before anything disappears.

Criteria for chart vs Kustomize
-------------------------------

A component **belongs in the chart** when all of the following are true:

* It is required to run Airflow (scheduler, API server, dag-processor,
  triggerer, workers).
* Removing it requires changes to Airflow's own configuration.

> * It has no external owner.

Reviewer: We also said pareto principle applies so would propose to add:
```suggestion
* It has no external owner.
* It is assumed that the larger majority (>80%) will need and use this function for productive use
```

A component **belongs in Kustomize** when any of the following are true:

* It can be expressed as a standalone Kubernetes resource without modifying
  chart-rendered resources.
* It is environment-specific (authentication schemes, logging backends,
  autoscaling controllers).
* It has an external owner (KEDA, Elasticsearch, any PostgreSQL distribution).

> * It requires CRDs that the chart does not install.

Reviewer: Similar here then:
```suggestion
* It requires CRDs that the chart does not install.
* It is used by only a few users such that additional complexity and maintenance burden does not pay out
```

If a component qualifies for Kustomize but no overlay exists yet, it stays in
the chart until the overlay is in place and verified.

Overlay structure
-----------------

Each overlay directory must contain:

* ``kustomization.yaml`` - the Kustomize entry point.
* The Kubernetes resources the overlay produces.

> * ``STATUS`` - a small YAML document declaring the verification state.

Reviewer: Should we make it a STATUS.YAML then if it is YAML content?

Author: Make sense!

Author: Changed :)

* ``README.rst`` - usage instructions and a migration guide from the
  equivalent chart-side configuration.

STATUS file format
------------------

> The ``STATUS`` file is a small YAML document with the following fields.

Reviewer: Smells and calls for a schema check like providers schema :-D (Can be another PR though)

Author: I agree, this should definitly need a schema definition and some level of check :D We can include to this one as well, let me check this out. Thanks Jens!

Author: I will make this in the next batch. Will resolve after adding it :)

Author: Added Pydantic validation to the same structure validation script, which is self-contained and should already be easy to maintain. This may not be the same with the provider schema, but it should work. Please let me know if

   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
   KIND, either express or implied.  See the License for the
   specific language governing permissions and limitations
   under the License.

Contributing Kustomize Overlays
===============================

This document is the authoritative reference for adding, evolving, and
retiring overlays under ``chart/kustomize-overlays/``.

Why this directory exists
-------------------------

The Airflow Helm chart has historically carried components that are not
Airflow-native. They make the chart heavier than it needs to be and pull
maintenance toward things that already have external owners. Expressing
these components as Kustomize overlays keeps the chart focused on Airflow
itself while still giving users a working starting point for the rest.

The chart never removes a component without a working overlay already in
place. Users always have a migration path before anything disappears.

Criteria for chart vs Kustomize
-------------------------------

A component **belongs in the chart** when all of the following are true:

* It is required to run Airflow (scheduler, API server, dag-processor,
  triggerer, workers).

> * Removing it requires changes to Airflow's own configuration.

Reviewer 2: ```suggestion
* Removing/adding it requires changes to Airflow's own configuration.
```
to include new features too

* It has no external owner.

> * It is assumed that the larger majority (>80%) will need and use this function for productive use

Reviewer 2: ```suggestion
* It is used by the larger majority of users (>80%) for production-ready environments
```
Maybe we should also determine it by the possible usage of the feature in terms of where it will be used (prod or non-prod env)? WDYT?

Author: Everything here can be used in production, right? I am not sure if we should make the environmental difference for them or think about it and give the user which environment to use on which level. How about like this?
```suggestion
* It is used by the larger majority of users (>80%) who will need and use this function for productive use
```

Reviewer 2: Looks good. I had in mind the case of the current `postgresql`, which was meant only for development, but as it was available for users, they just used it. I fully agree that there should not be a distinction between environments, and if something is added, we assume that it can be used for dev or prod environments.

Reviewer: Ah you were more nit-picky on the wording that I prepared the proposal :-D Yes of course can and should be used in dev/test as well as production! I wanted to express rather: 80% need it. (then if they need it probably in all stages)

So in short it can be:

```suggestion
* It is used by the larger majority of users (>80%)
```

Reviewer 2: Agreed

A component **belongs in Kustomize** when any of the following are true:

* It can be expressed as a standalone Kubernetes resource without modifying
  chart-rendered resources.
* It is environment-specific (authentication schemes, logging backends,
  autoscaling controllers).

> * It has an external owner (KEDA, Elasticsearch, any PostgreSQL distribution).

Reviewer 2: ```suggestion
  autoscaling controllers, etc.).
* It has an external owner (KEDA, Elasticsearch, any PostgreSQL distribution, etc.).
```

* It requires CRDs that the chart does not install.

> * It is used by only a few users such that additional complexity and maintenance burden does not pay out

Reviewer 2: ```suggestion
* It is used by a minority of users, such that the additional complexity and maintenance burden do not pay off
```
I think that sounds a little better
